Scripts and data update

Arun-bioinformatics opened the pull request:
modified:   CEN_multiqc_config_v2.1.0.yaml
- New key added inside table_cond_formatting_rules to test records that is only found in read data

**new file:**   idUniqueIdRelationship.json
- Essential file to describe relationship between IDs found in config.yaml and multiqc.json

**new file:**   mypackage/functions.py
- Functions for QC classifier project

**new file:**   codeInfo.py
- Non essential python script which explains the applications of functions from mypackage

**new file:**   fullAssembly.py
- Script which unifies functions from mypackage to ouput potential .json

**new file:** CEN_multiqc_config_v2.1.0.yaml
- Included config file that will be used as example for the QC classifier project.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/eastgenomics/QC_Classifier/2)
<!-- Reviewable:end -->

Commit:
new file:   CEN_multiqc_config_v2.1.0.yaml

Included config file that will be used as example for the QC classifier
project.

Commit:
Initial commits for QC projects

	modified:   CEN_multiqc_config_v2.1.0.yaml
        - New key added inside table_cond_formatting_rules to test records that is only found in read data

    new file:   idUniqueIdRelationship.json
        - Essential file to describe relationship between IDs found in config.yaml and multiqc.json
	new file:   mypackage/functions.py
        - Functions for QC classifier project
    new file:   codeInfo.py
        - Non essential python script which explains the applications of functions from mypackage
	new file:   fullAssembly.py
        - Script which unifies functions from mypackage to ouput potential .json

Commit:
new file:   multiqc_data.json

Includes multiqc_data.json used as an example.

jethror1 reviewed (changes requested):
> Reviewed 1 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 1 of 6 files reviewed, 28 unresolved discussions (waiting on @Arun-bioinformatics)

___
*[`CEN_multiqc_config_v2.1.0.yaml` line 1 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5yWSN3xYRdcsnyG3S:-Nd5yWSN3xYRdcsnyG3T:b-e876l9) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/CEN_multiqc_config_v2.1.0.yaml#L1)):*
> ```YAML
> ###################################################
> ```

not sure if this file needs to be included, unless you want to keep it in a test data dir or similar?
___
*[`codeInfo.py` line 1 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5yfUh3qzFjOd5e4QO:-Nd5yfUh3qzFjOd5e4QP:b-wlgklm) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/codeInfo.py#L1)):*
> ```Python
> import mypackage.functions as Classifier
> ```

is this just for your own notes? if so can you remove it (or I'd add it in a markdown file or something called `notes.md` or something to make it clear its not a real script)
___
*[`fullAssembly.py` line 2 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62W-C1cZQJFKSB_SO:-Nd62W-C1cZQJFKSB_SP:bbeppyf) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L2)):*
> ```Python
> import mypackage.functions as Classifier
> from ruamel.yaml import YAML
> ```

any reason for using this YAML parser? afaik pyyaml is the most widely used / oldest (and should just be in the standard library): https://pypi.org/project/PyYAML/



_Code quote:_
```Python
from ruamel.yaml import YAML
```
___
*[`fullAssembly.py` line 3 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5zBh83NXzClN1D7Z7:-Nd5zBh83NXzClN1D7Z8:b5t5mig) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L3)):*
> ```Python
> import mypackage.functions as Classifier
> from ruamel.yaml import YAML
> import json
> ```

convention for ordering imports is alphanumerically following the order:
1.  standard library imports
2.  related third party imports
3.  local application/library specific imports

i.e. for yours:
```
import json

from ruamel.yaml import YAML

import mypackage.functions as Classifier
```

a linter such as pylint enabled in vscode should highlight this for you, here's an example with a load of imports: https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/master/resources/home/dnanexus/generate_workbook/utils/excel.py#L1



_Code quote:_
```Python
import mypackage.functions as Classifier
from ruamel.yaml import YAML
import json
```
___
*[`fullAssembly.py` line 26 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5ym0U-hzgQhJ0uxnm:-Nd5ym0U-hzgQhJ0uxnn:b-xnsaku) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L26)):*
> ```Python
> 
> 
> '''
> ```

if this is going to stay as a file docstring it should be at the top above the imports



_Code quote:_
```Python
'''
We have all the functions needed to generate the individual
outputs that wee need, all we have to do is join everything together in
a complex dictionary.

All inputs:
    - idUniqueIDRelationship.json
    - multiqc_data.json
    - .yaml file

Output:
We will add more complexity as we go through it but starting with the following
    {"record":["sample_ids":{"Value":1.0, "Status":"pass"}]}

Next step of complexity
    {"thresholds": parameters, "record":["sample_ids":{"Value":1.0, "Status":"pass"}]}

Future next step of complexity
    {Header_ID: {"thresholds": parameters, "record":["sample_ids":{"Value":1.0, "Status":"pass"}]}}


'''
```
___
*[`fullAssembly.py` line 29 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd61CjoCI1oRPvQ13o2:-Nd61CjoCI1oRPvQ13o3:bvapl0f) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L29)):*
> ```Python
> 
> # List of samples
> multiqc_data = json.load(open("multiqc_data.json"))
> ```

reading a file like this with `open()` will leave an open pointer to the file and keep it open, the better way to read from files is:
```
with open(some_file) as f:
    data = json.load(f)
```

when the file has been read in and drops out of the `with` statement it will automatically close the file



_Code quote:_
```Python
multiqc_data = json.load(open("multiqc_data.json"))
```
___
*[`fullAssembly.py` line 32 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oJyA7CgGyfJhtbpS:-Nd9oJyB1j7nNpbe_EPP:b-wxf9qa) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L32)):*
> ```Python
> sample_lists = Classifier.getSampleLists(multiqc_data)
> 
> # List of uniqueIDs
> ```

are these unique_ids the ids of fields we have set in the config? if so calling it 'config_fields' might be clearer, as when you're dealing with samples using ID gets confusing



_Code quote:_
```Python
# List of uniqueIDs
```
___
*[`fullAssembly.py` line 35 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9pBuD7epKDAG2QsVV:-Nd9pBuD7epKDAG2QsVW:b-2p3o5i) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L35)):*
> ```Python
> yaml_file = "CEN_multiqc_config_v2.1.0.yaml"
> with open(yaml_file,'r') as file:
>     yaml_content = YAML().load(file)
> ```

I would have this as its own `read_config()` function in a utils file or similar and call it, example of similar: https://github.com/eastgenomics/optimised_filtering/blob/main/utils/file_utils.py



_Code quote:_
```Python
yaml_file = "CEN_multiqc_config_v2.1.0.yaml"
with open(yaml_file,'r') as file:
    yaml_content = YAML().load(file)
```
___
*[`fullAssembly.py` line 52 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9onA6CA4m3fbQAk-e:-Nd9onA74kLZVwOQ7Ex9:b-xvu0t9) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L52)):*
> ```Python
>         record_structure = None
> 
>         parameters = Classifier.getUniqueParameters(uniqueID, "CEN_multiqc_config_v2.1.0.yaml")
> ```

 I assume the config wants to be an input to the script and not hardcoded, I'd recommend looking at parse_args if you haven't before, here's a couple of examples :
- https://github.com/eastgenomics/optimised_filtering/blob/1b75fe42abe0f5733eeff4da70087af3c7d13797/add_optimised_filtering.py#L11
- https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/4b914db6963b179b31bc28f93b24c7c889b6445f/resources/home/dnanexus/generate_workbook/generate_workbook.py#L64 



_Code quote:_
```Python
"CEN_multiqc_config_v2.1.0.yaml"
```
___
*[`fullAssembly.py` line 58 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oazB3NI5WDfQGiO5:-Nd9oazB3NI5WDfQGiO6:b-jjfhsz) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L58)):*
> ```Python
>         # Parameters values are around the 100s
>         # Whereas associated values are around 1.0-0.9s
>         if uniqueID == 'mqc-generalstats-picard-PCT_TARGET_BASES_20X':
> ```

is this the only field where this happens?



_Code quote:_
```Python
'mqc-generalstats-picard-PCT_TARGET_BASES_20X':
```
___
*[`fullAssembly.py` line 72 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9o6qCBzvUjRB-Pe_d:-Nd9o6qCBzvUjRB-Pe_e:bb7phvt) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L72)):*
> ```Python
>             record_structure = []
> 
>             # If value found, then retrieve value and status for all elements in tupple
> ```

typo



_Code quote:_
```Python
tupple
```
___
*[`fullAssembly.py` line 89 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9nJHA6QQbaixjEaFv:-Nd9nJHA6QQbaixjEaFw:b-61cfwf) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L89)):*
> ```Python
>             print(messageFromClassifier)
> ```

convention here if this is your main entrypoint script is to format it like this and not to have code 'bare' outside of functions, as this will make everything in the global scope which is not the best, here's an example: https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/4b914db6963b179b31bc28f93b24c7c889b6445f/resources/home/dnanexus/generate_workbook/generate_workbook.py#L511

if its a tiny script with not much 'main' logic then it can be fine to put everything under the `if __name__ == "__main__":` statement, but I generally like to chuck it all in a `main()` function for readability
___
*[`mypackage/functions.py` line 3 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd60o2s-bi0Qe1JQcmu:-Nd60o2s-bi0Qe1JQcmv:b-85eh17) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L3)):*
> ```Python
> import subprocess
> import json
> from ruamel.yaml import YAML
> ```

import ordering as above



_Code quote:_
```Python
import subprocess
import json
from ruamel.yaml import YAML
```
___
*[`mypackage/functions.py` line 4 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd60Xwa4es-5LgroqTb:-Nd60Xwa4es-5LgroqTc:by88c13) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L4)):*
> ```Python
> import json
> from ruamel.yaml import YAML
> ```

I would recommend renaming this file to something more descriptive like `utils.py`, and probably the directory either as `bin/` or `utils/` 
___
*[`mypackage/functions.py` line 6 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62uDH8x7WNsj4fDu1:-Nd62uDH8x7WNsj4fDu2:b-d7bxid) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L6)):*
> ```Python
> 
> 
> def getHeaderID(unique_id):
> ```

function naming should follow camel case: i.e. `def get_header_id()`



_Code quote:_
```Python
def getHeaderID(unique_id):
```
___
*[`mypackage/functions.py` line 24 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd635b8CSmzgnvAnmcV:-Nd635b8CSmzgnvAnmcW:b-mrbhwq) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L24)):*
> ```Python
>     cmd = f'jq -r --arg UNIQUE "{unique_id}" \'first(.[] | select(.Unique_ID == $UNIQUE) | .Header_ID)\' idUniqueIdRelationship.json'
>     header_id = subprocess.check_output(cmd, shell=True).decode().strip()
>     return header_id
> ```

it would be better to stick to one language where you can instead of using subprocess to call others (except in cases where you want to call bcftools etc), the json parser in python is pretty good - let me know if you want help writing something to parse it



_Code quote:_
```Python
    # Use jq to extract header ID based on the current unique ID
    cmd = f'jq -r --arg UNIQUE "{unique_id}" \'first(.[] | select(.Unique_ID == $UNIQUE) | .Header_ID)\' idUniqueIdRelationship.json'
    header_id = subprocess.check_output(cmd, shell=True).decode().strip()
    return header_id
```
___
*[`mypackage/functions.py` line 41 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd63sqg8xA6BnDFZgrC:-Nd63sqhE2u5n8IW8165:b-deaqj6) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L41)):*
> ```Python
> 
>     with open(yaml_file, 'r') as file:
>         yaml_content = YAML().load(file)
> ```

if you're going to read this file more times to extract things, consider moving this to the main function where this function is called and pass in the yaml data as an input



_Code quote:_
```Python
    with open(yaml_file, 'r') as file:
        yaml_content = YAML().load(file)
```
___
*[`mypackage/functions.py` line 48 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd68KYdAYzStFNTBu5o:-Nd68KYdAYzStFNTBu5p:bn5ozdi) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L48)):*
> ```Python
> 
> def getSampleLists(multiqc_data):
>     '''
> ```

I would consider using the samplesheet as an input to use as truth for getting the sample IDs from instead of trying to parse from the report sections since the samplesheet is the truth source for samplenames on a run, here's a function you can steal to parse it: https://github.com/eastgenomics/eggd_conductor/blob/262d9abcd20bf141a92d9f59fe3498cfe48d0dc7/resources/home/dnanexus/run_workflows/run_workflows.py#L29
___
*[`mypackage/functions.py` line 87 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9iw1F2jws_HnkVJ6H:-Nd9iw1F2jws_HnkVJ6I:b-crqjc4) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L87)):*
> ```Python
> def getControlLists(multiqc_data):
>     '''
>     Creates a structured list of controls used in the MultiQC run
> ```

by control do you mean the NA12878 samples etc? for this we use sompy in Uranus for the haemonc samples, but not sure if there's an equivalent in Helios/TSO500. We can pick controls up through the 2nd field name having a Q in if its anything from Epic, why do you need to pick up the control specifically?



_Code quote:_
```Python
controls
```
___
*[`mypackage/functions.py` line 93 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9jQzw9DrzJey6UmcP:-Nd9jQzw9DrzJey6UmcQ:bd0pwnr) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L93)):*
> ```Python
> 
>     Output: 
>         ([controlSample_snp_all, controlSample_snp_pass], [controlSample_indel_all, controlSample_indel_pass])
> ```

inputs and outputs could do with more detail (like you have for some of the others), it really helps for future you to come back to if you try describe what/where the data is from/for and include the type, i.e. I like:
```
Parameters
----------------
multiqc_data : dict
     dict of data from multiqc_data.json to parse from
```

Thats the numpy/pandas docstring style, but feel free to use Googles or whatever you prefer (if you're using the autodoc extension for vscode you can set this in the preferences), some more examples of different peoples code to look at:
- https://github.com/eastgenomics/trendyQC/blob/local_dev/trendyqc/trend_monitoring/management/commands/_dnanexus_utils.py
- https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/master/resources/home/dnanexus/generate_workbook/utils/columns.py
- https://github.com/eastgenomics/optimised_filtering/blob/main/utils/vcf.py



_Code quote:_
```Python
    Input:
        multiqc_data

    Output:
        ([controlSample_snp_all, controlSample_snp_pass], [controlSample_indel_all, controlSample_indel_pass])
```
___
*[`mypackage/functions.py` line 119 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9l5l18Vji3HrLjMPj:-Nd9l5l18Vji3HrLjMPk:b-98dm3r) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L119)):*
> ```Python
>     data = {}
> 
>     # If and elif statements only are applicable when given ids from getControlLists
> ```

so here are you only returning data for happy for controls and then everything else for the other samples? again need to think about controls not just being RD/happy 



_Code quote:_
```Python
 If and elif statemen
```
___
*[`mypackage/functions.py` line 141 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9lNmT2riIyv5yi0ym:-Nd9lNmT2riIyv5yi0yn:b-rj3skn) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L141)):*
> ```Python
> def getStatus(value, parameters):
>     '''
>     Function to determine the appropriate status based on given value and parameters.
> ```

by status do you mean if its passed/failed QC thresholds?



_Code quote:_
```Python
 appropriate status 
```
___
*[`mypackage/functions.py` line 161 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9la755Wn2puOiQCwD:-Nd9la755Wn2puOiQCwE:b2q0ytd) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L161)):*
this is the same as doing just  `if possible_status` as anything not False/None/an empty object will evaluate to true:
```
>>> a='foo'
>>> b=True
>>> c={'some': 'dict'}
>>> if a: print('y')
... 
y
>>> if b: print('y')
... 
y
>>> if c: print('y')
... 
y
```



_Code quote:_
```Python
if possible_status == True:
```
___
*[`mypackage/functions.py` line 163 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9mLg7Ba3eNY52hfHh:-Nd9mLg7Ba3eNY52hfHi:bsbphs8) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L163)):*
> ```Python
>         if possible_status == True:
>             # If it is a boolean == True, check if value is a string == "true"
>             # *IMPORTANT*, DISCUSS CAVEATS FOR FALSE
> ```

things you need to remember I would comment as `# TODO: ` as these get highlighted in the code and people know to look for them



_Code quote:_
```Python
# *IMPORTANT*, DISCUSS CAVEATS FOR FALSE
```
___
*[`mypackage/functions.py` line 164 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9mDtE-3zHsYqIphfv:-Nd9mDtE-3zHsYqIphfw:b-sn3sej) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L164)):*
> ```Python
>             # If it is a boolean == True, check if value is a string == "true"
>             # *IMPORTANT*, DISCUSS CAVEATS FOR FALSE
>             if value == "true":
> ```

here does value stored as a string come from and why are you storing True as a string?



_Code quote:_
```Python
if value == "true":
```
___
*[`mypackage/functions.py` line 166 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9m3kx2LHC2-BohWSt:-Nd9m3kx2LHC2-BohWSu:b-7as0x8) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L166)):*
> ```Python
>             if value == "true":
>                 status = "pass"
>                 continue
> ```

continue statement here is not needed as there's nothing further in the loop



_Code quote:_
```Python
continue
```
___
*[`mypackage/functions.py` line 175 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9mZe80Q8wk8Lk4TPR:-Nd9mZe80Q8wk8Lk4TPS:bb7phvt) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L175)):*
> ```Python
> 
>             # Checking if any of the following conditions exist
>             # Or statements are necessary to prevent any condition with value 0 to be treted as false 
> ```

typo



_Code quote:_
```Python
treted 
```
___
*[`mypackage/functions.py` line 176 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9n2o6CxsiT7HrHGCi:-Nd9n2o6CxsiT7HrHGCj:b1f8aek) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L176)):*
> ```Python
>             # Checking if any of the following conditions exist
>             # Or statements are necessary to prevent any condition with value 0 to be treted as false 
>             if condition.get('gt') or condition.get('gt') == 0:
> ```

what does the structure of `conditions` look like here?



_Code quote:_
```Python
if condition.get('gt') or condition.get('gt') == 0:
```


<!-- Sent from Reviewable.io -->

Commit:
Minor changes
	renamed:    mypackage/functions.py -> bin/utils.py
    - "Unique_ID" variable changed to "config_field".
    - Converted all functions from camelCase to snake_case.
    - updated yaml module to pyyaml.
    - included read_config() function
	modified:   fullAssembly.py
	deleted:    multiqc_data.json
	renamed:    codeInfo.py -> notes.md
	renamed:    idUniqueIdRelationship.json -> resources/idUniqueIdRelationship.json
	renamed:    CEN_multiqc_config_v2.1.0.yaml -> testfiles/CEN_multiqc_config_v2.1.0.yaml

Commit:
Bugfix for getting parameters
	modified:   bin/utils.py
	modified:   fullAssembly.py
    - Removal of continue statements, tested without them and works well.

Arun-bioinformatics reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 0 of 7 files reviewed, 26 unresolved discussions (waiting on @jethror1)

___
*[`fullAssembly.py` line 2 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62W-C1cZQJFKSB_SO:-NdGSgoEDlcmPahmdEAr:b-59dpow) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L2)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

any reason for using this YAML parser? afaik pyyaml is the most widely used / oldest (and should just be in the standard library): https://pypi.org/project/PyYAML/
</blockquote></details>

Cool, now I am using PyYAML version 6.0.1
___
*[`fullAssembly.py` line 3 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5zBh83NXzClN1D7Z7:-NdUc-Sk9XZM1zhQ0-B5:b-3j0b44) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L3)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

convention for ordering imports is alphanumerically following the order:
1.  standard library imports
2.  related third party imports
3.  local application/library specific imports

i.e. for yours:
```
import json

from ruamel.yaml import YAML

import mypackage.functions as Classifier
```

a linter such as pylint enabled in vscode should highlight this for you, here's an example with a load of imports: https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/master/resources/home/dnanexus/generate_workbook/utils/excel.py#L1
</blockquote></details>

The order of imports was modified to the following:

```
import json
import yaml
import mypackage.functions as Classifier
```
___
*[`fullAssembly.py` line 26 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5ym0U-hzgQhJ0uxnm:-NdUcHOw5LzO_jrBwPAr:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L26)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

if this is going to stay as a file docstring it should be at the top above the imports
</blockquote></details>

Done.
___
*[`fullAssembly.py` line 29 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd61CjoCI1oRPvQ13o2:-NdUq3KJEjEK8nuv4FSQ:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L29)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

reading a file like this with `open()` will leave an open pointer to the file and keep it open, the better way to read from files is:
```
with open(some_file) as f:
    data = json.load(f)
```

when the file has been read in and drops out of the `with` statement it will automatically close the file
</blockquote></details>

Done.
___
*[`fullAssembly.py` line 32 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oJyA7CgGyfJhtbpS:-NdUrQBt0d181b5D52ma:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L32)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

are these unique_ids the ids of fields we have set in the config? if so calling it 'config_fields' might be clearer, as when you're dealing with samples using ID gets confusing
</blockquote></details>

Done.
___
*[`fullAssembly.py` line 35 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9pBuD7epKDAG2QsVV:-NdUtySK1Go8tMEjCQSx:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L35)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I would have this as its own `read_config()` function in a utils file or similar and call it, example of similar: https://github.com/eastgenomics/optimised_filtering/blob/main/utils/file_utils.py
</blockquote></details>

Done.
___
*[`fullAssembly.py` line 58 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oazB3NI5WDfQGiO5:-NdUy9FZEThr8oi8WRA6:b-x738vw) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L58)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

is this the only field where this happens?
</blockquote></details>

Yes, I am not sure of how to code it in a better way, but I am open to new suggestions.
___
*[`fullAssembly.py` line 72 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9o6qCBzvUjRB-Pe_d:-NdV383-Bd48t_-PR1YN:bvm7z8d) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L72)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Resolved
___
*[`codeInfo.py` line 1 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5yfUh3qzFjOd5e4QO:-NdGNqvWBU2CVpVaQ7Oh:b-zb7077) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/codeInfo.py#L1)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

is this just for your own notes? if so can you remove it (or I'd add it in a markdown file or something called `notes.md` or something to make it clear its not a real script)
</blockquote></details>

This will now be added onto a `notes.md` file
___
*[`CEN_multiqc_config_v2.1.0.yaml` line 1 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd5yWSN3xYRdcsnyG3S:-NdG45MEClzL766z5Gs3:b3ehwen) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/CEN_multiqc_config_v2.1.0.yaml#L1)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

not sure if this file needs to be included, unless you want to keep it in a test data dir or similar?
</blockquote></details>

Would want to keep it in a test_files directory, since current Python scripts are getting built while using the config file. Will be removed in the future.
___
*[`mypackage/functions.py` line 3 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd60o2s-bi0Qe1JQcmu:-NdV3Zyw32eolbn3o5c2:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L3)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

import ordering as above
</blockquote></details>

Done.
___
*[`mypackage/functions.py` line 4 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd60Xwa4es-5LgroqTb:-NdV3c7s2SnF5V-uclyd:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L4)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I would recommend renaming this file to something more descriptive like `utils.py`, and probably the directory either as `bin/` or `utils/`
</blockquote></details>

Done.
___
*[`mypackage/functions.py` line 6 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62uDH8x7WNsj4fDu1:-NdG9RZt4KKogl2w2kPs:be6wks8) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L6)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

function naming should follow camel case: i.e. `def get_header_id()`
</blockquote></details>

Will change all functions to the same format like `def map_header_id()`, but wouldn't that be called snake case?
___
*[`mypackage/functions.py` line 24 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd635b8CSmzgnvAnmcV:-NdVW_3N6n5GdAJgr-UO:b-vtmuvm) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L24)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

it would be better to stick to one language where you can instead of using subprocess to call others (except in cases where you want to call bcftools etc), the json parser in python is pretty good - let me know if you want help writing something to parse it
</blockquote></details>

Function is named as `map_header_id()`. The function uses  python's json module instead.
___
*[`mypackage/functions.py` line 41 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd63sqg8xA6BnDFZgrC:-NdVXPUQ0LB3NWCFftzx:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L41)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

if you're going to read this file more times to extract things, consider moving this to the main function where this function is called and pass in the yaml data as an input
</blockquote></details>

Done.
___
*[`mypackage/functions.py` line 87 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9iw1F2jws_HnkVJ6H:-NdVY3Ya2usIiwewwVK3:b37vv74) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L87)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

by control do you mean the NA12878 samples etc? for this we use sompy in Uranus for the haemonc samples, but not sure if there's an equivalent in Helios/TSO500. We can pick controls up through the 2nd field name having a Q in if its anything from Epic, why do you need to pick up the control specifically?
</blockquote></details>

Not really, I mean this as controls:
![image.png](https://files.reviewableusercontent.io/images/reviewable/75141216/qCFQFi7TRBe7Has8Lbesc83R/image.png)
___
*[`mypackage/functions.py` line 93 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9jQzw9DrzJey6UmcP:-NdV_dinDt4TclLzoAwo:bqfdg2o) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L93)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

inputs and outputs could do with more detail (like you have for some of the others), it really helps for future you to come back to if you try describe what/where the data is from/for and include the type, i.e. I like:
```
Parameters
----------------
multiqc_data : dict
     dict of data from multiqc_data.json to parse from
```

Thats the numpy/pandas docstring style, but feel free to use Googles or whatever you prefer (if you're using the autodoc extension for vscode you can set this in the preferences), some more examples of different peoples code to look at:
- https://github.com/eastgenomics/trendyQC/blob/local_dev/trendyqc/trend_monitoring/management/commands/_dnanexus_utils.py
- https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/master/resources/home/dnanexus/generate_workbook/utils/columns.py
- https://github.com/eastgenomics/optimised_filtering/blob/main/utils/vcf.py
</blockquote></details>

Done, Inputs and outputs explained in more detail in the next update.
___
*[`mypackage/functions.py` line 141 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9lNmT2riIyv5yi0ym:-NdVaBfz7IJj23JcLOkC:b-exlrfs) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L141)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

by status do you mean if its passed/failed QC thresholds?
</blockquote></details>

Yes, clarified this in next update
___
*[`mypackage/functions.py` line 161 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9la755Wn2puOiQCwD:-NdVcMqs-sVa2L6qYoav:bcwkplu) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L161)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

this is the same as doing just  `if possible_status` as anything not False/None/an empty object will evaluate to true:
```
>>> a='foo'
>>> b=True
>>> c={'some': 'dict'}
>>> if a: print('y')
... 
y
>>> if b: print('y')
... 
y
>>> if c: print('y')
... 
y
```
</blockquote></details>

Yes, I wanted to check whether the possible status was a boolean and equal to True, I have changed this to the following

``` 
if isinstance(possible_status, bool) and possible_status is True:
````
___
*[`mypackage/functions.py` line 163 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9mLg7Ba3eNY52hfHh:-NdVby2EAX8jaDXHAjiw:bn5fyq) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L163)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

things you need to remember I would comment as `# TODO: ` as these get highlighted in the code and people know to look for them
</blockquote></details>

Done. Comment changed to include `# TODO:`
___
*[`mypackage/functions.py` line 164 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9mDtE-3zHsYqIphfv:-NdVcqe-A1YNz2dccqeO:b-lmcv5e) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L164)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

here does value stored as a string come from and why are you storing True as a string?
</blockquote></details>


This is only applied on one single case where the config field `Match_Sexes` has the following parameters in the config table:

```
true:
    - s_eq: 'pass'
fail:
    - s_eq: 'fail'
warn:
    - s_eq: 'NA'
```
But the value for each sample does not return neither of these strings, but returns either a "true" "false" string instead, therefore the recorded status for all of these would be assigned as 'unknown'. This if statement supersedes given conditions and automatically assigns status "pass" if returned value is a string equal to "true". 
___
*[`mypackage/functions.py` line 166 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9m3kx2LHC2-BohWSt:-NdVd-mF8UM_qWjEx1qm:b-4xzhyx) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L166)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

continue statement here is not needed as there's nothing further in the loop
</blockquote></details>

I see, it works the same way when continue statement is not included
___
*[`mypackage/functions.py` line 176 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9n2o6CxsiT7HrHGCi:-NdVdwm63icYf92Vdvyk:b7txhnv) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L176)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

what does the structure of `conditions` look like here?
</blockquote></details>

For each possible status (pass/warn/fail) there can is a list of 1-4 conditions, the keys for these conditions can be one of the following (gt, lt, eq, s_eq).
TODO: not sure of what the question is here, need to ask for more info with what do I have to do


<!-- Sent from Reviewable.io -->

jethror1 reviewed (changes requested):
> Reviewed 5 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 5 of 7 files reviewed, 9 unresolved discussions (waiting on @Arun-bioinformatics)

___
*[`fullAssembly.py` line 58 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oazB3NI5WDfQGiO5:-Nd_T2Mw7oOde8iPZq1A:bd7l74p) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L58)):*
<details><summary><i>Previously, Arun-bioinformatics (Arun Karnani Khemlani) wrote…</i></summary><blockquote>

Yes, I am not sure of how to code it in a better way, but I am open to new suggestions.
</blockquote></details>

nope no real issue, just if there were going to be more I'd change it but this is sensible
___
*[`bin/utils.py` line 24 at r3](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd_TPQDCMJm84yZ3Lrs:-Nd_TPQDCMJm84yZ3Lrt:b-e7p4g) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ebd119db2aa43713a0a76e8dd33bafb179b72546/bin/utils.py#L24)):*
> ```Python
>     """
>     # First get file from resources/ folder
>     with open("gitRepo/QC_Classifier/resources/idUniqueIdRelationship.json",
> ```

personally if you're doing paths relative to your script (i.e in a different directory but in the same code base, I would use something like `pathlib` as its more robust, and reference the relative folder, should be something like this:
```
from pathlib import Path

Path(__file__).parent.joinpath('../resources/idUniqueIdRelationship.json')
```

docs for pathlib are pretty good too and let you write code that will work on linux and windows: https://docs.python.org/3/library/pathlib.html



_Code quote:_
```Python
gitRepo/QC_Classifier/resources/idUniqueIdRelationship.json
```
___
*[`mypackage/functions.py` line 6 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62uDH8x7WNsj4fDu1:-Nd_TBsJElkJLlJ_fGor:b8wmwzm) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L6)):*
<details><summary><i>Previously, Arun-bioinformatics (Arun Karnani Khemlani) wrote…</i></summary><blockquote>

Will change all functions to the same format like `def map_header_id()`, but wouldn't that be called snake case?
</blockquote></details>

yep sorry, end of day brain not working


<!-- Sent from Reviewable.io -->

jethror1 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 5 of 7 files reviewed, 10 unresolved discussions (waiting on @Arun-bioinformatics)

___
*[`fullAssembly.py` line 3 at r2](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd_cYj45lNKWAS0tznQ:-Nd_cYj45lNKWAS0tznR:bb7phvt) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/d4eed60745b0c23e682ff1a2d73fc3aecb0cacd5/fullAssembly.py#L3)):*
> ```Python
> """
> We have all the functions needed to generate the individual 
> outputs that wee need, all we have to do is join everything together in
> ```

typo



_Code quote:_
```Python
wee
```


<!-- Sent from Reviewable.io -->

Commit:
Ignore multiqc_data.json

Commit:
Added: Pathlib and Argparse
	modified:   bin/utils.py
    - Included module pathlib
	renamed:    fullAssembly.py -> full_assembly.py
    - Included argparse module to parse data and config file into
      this python script

Arun-bioinformatics reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 5 of 8 files reviewed, 9 unresolved discussions (waiting on @jethror1)

___
*[`fullAssembly.py` line 52 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9onA6CA4m3fbQAk-e:-Nd_icOHCZKP-xBcnYn7:b-my89n3) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L52)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I assume the config wants to be an input to the script and not hardcoded, I'd recommend looking at parse_args if you haven't before, here's a couple of examples :
- https://github.com/eastgenomics/optimised_filtering/blob/1b75fe42abe0f5733eeff4da70087af3c7d13797/add_optimised_filtering.py#L11
- https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/4b914db6963b179b31bc28f93b24c7c889b6445f/resources/home/dnanexus/generate_workbook/generate_workbook.py#L64
</blockquote></details>

Included argparse module to parse the multiqc_data.json and config.yaml
___
*[`fullAssembly.py` line 58 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9oazB3NI5WDfQGiO5:-Nd_dTCO8H5Jw8jfQcEw:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L58)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

nope no real issue, just if there were going to be more I'd change it but this is sensible
</blockquote></details>

Done.
___
*[`fullAssembly.py` line 3 at r2](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd_cYj45lNKWAS0tznQ:-Nd_dFzH3mk8JRpxG64Z:bb6un1j) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/d4eed60745b0c23e682ff1a2d73fc3aecb0cacd5/fullAssembly.py#L3)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done
___
*[`bin/utils.py` line 24 at r3](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd_TPQDCMJm84yZ3Lrs:-Nd_jQ5FAzmu6IyQ9Pdk:b1mw53) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ebd119db2aa43713a0a76e8dd33bafb179b72546/bin/utils.py#L24)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

personally if you're doing paths relative to your script (i.e in a different directory but in the same code base, I would use something like `pathlib` as its more robust, and reference the relative folder, should be something like this:
```
from pathlib import Path

Path(__file__).parent.joinpath('../resources/idUniqueIdRelationship.json')
```

docs for pathlib are pretty good too and let you write code that will work on linux and windows: https://docs.python.org/3/library/pathlib.html
</blockquote></details>

Thank you! that is really helpful to know.
___
*[`mypackage/functions.py` line 6 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd62uDH8x7WNsj4fDu1:-Nd_dZv151ViSyJwkWo8:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L6)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

yep sorry, end of day brain not working
</blockquote></details>

Done.
___
*[`mypackage/functions.py` line 119 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9l5l18Vji3HrLjMPj:-Nd_uZQQ7gkDCs5Dv_3O:bx67f66) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L119)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

so here are you only returning data for happy for controls and then everything else for the other samples? again need to think about controls not just being RD/happy
</blockquote></details>

Need to discuss about it in person


<!-- Sent from Reviewable.io -->

Commit:
Included SampleSheet.csv and description in README.md
	modified:   README.md
	new file:   testfiles/SampleSheet.csv

Commit:
Modified bin/utils.py & full_assembly.py

    modified:   bin/utils.py
    - Modified `get sample_lists()` to include SampleSheet.csv as input
    - BUGFIX: `get_status()` outputs "fail" instead of "unknown" when input value is "false"
	modified:   full_assembly.py
    - Updated doctring to include json schema output structure
    - Include `SampleSheet` in argparse
    - Chunk main script into `main()`.

Arun-bioinformatics reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 5 of 10 files reviewed, 9 unresolved discussions (waiting on @jethror1)

___
*[`fullAssembly.py` line 89 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd9nJHA6QQbaixjEaFv:-NdjnABH0WU92aBvX8tx:b-896fix) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/fullAssembly.py#L89)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

convention here if this is your main entrypoint script is to format it like this and not to have code 'bare' outside of functions, as this will make everything in the global scope which is not the best, here's an example: https://github.com/eastgenomics/eggd_generate_variant_workbook/blob/4b914db6963b179b31bc28f93b24c7c889b6445f/resources/home/dnanexus/generate_workbook/generate_workbook.py#L511

if its a tiny script with not much 'main' logic then it can be fine to put everything under the `if __name__ == "__main__":` statement, but I generally like to chuck it all in a `main()` function for readability
</blockquote></details>

Done.
___
*[`mypackage/functions.py` line 48 at r1](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-Nd68KYdAYzStFNTBu5o:-Nddh7dnDCqj_RdpYdv2:brm5ab8) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ef6521e29db2692ddceff55635e2baeb3b413282/mypackage/functions.py#L48)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

I would consider using the samplesheet as an input to use as truth for getting the sample IDs from instead of trying to parse from the report sections since the samplesheet is the truth source for samplenames on a run, here's a function you can steal to parse it: https://github.com/eastgenomics/eggd_conductor/blob/262d9abcd20bf141a92d9f59fe3498cfe48d0dc7/resources/home/dnanexus/run_workflows/run_workflows.py#L29
</blockquote></details>

Spoke to Matt about this, and he seems to agree.

Take for example one sample with sampleID '123991777-23187R0015-23NGCEN9-9527-F-99347387', this ID would be found in a tuple like so:

('123991777-23187R0015-23NGCEN9-9527-F-99347387', <---- right here
('123991777-23187R0015-23NGCEN9-9527-F-99347387_S2_L001_R2',
'123991777-23187R0015-23NGCEN9-9527-F-99347387_S2_L002_R1',
'123991777-23187R0015-23NGCEN9-9527-F-99347387_S2_L002_R2',
'123991777-23187R0015-23NGCEN9-9527-F-99347387_S2_L001_R1'))

SampleSheet.csv does include Sample IDs, but it not include read IDs (like '123991777-23187R0015-23NGCEN9-9527-F-99347387_S2_L001_R2').  So will get read IDs from multiqc_data.json file, open to other suggestions.


<!-- Sent from Reviewable.io -->

jethror1 reviewed (changes requested):
> Reviewed 1 of 3 files at r4, 2 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 8 of 10 files reviewed, 5 unresolved discussions (waiting on @Arun-bioinformatics)

___
*[`full_assembly.py` line 40 at r5](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-NdkC56kEiuDoQN0YhnO:-NdkC56kEiuDoQN0YhnP:b-siglec) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ccd8c66bc9ff29ccf0ffac4ff4079f07eaf56f5f/full_assembly.py#L40)):*
> ```Python
>                     help='filepath to config.yaml file')
> 
> args = parser.parse_args()
> ```

please can this be in its own function too please and not global



_Code quote:_
```Python
parser = argparse.ArgumentParser(description='QC Classifier for multiqc data')
parser.add_argument('samplesheet', type=str,
                    help='Filepath to SampleSheet.csv file')
parser.add_argument('data', type=str,
                    help='filepath to multiqc.json file')
parser.add_argument('config', type=str,
                    help='filepath to config.yaml file')

args = parser.parse_args()
```
___
*[`full_assembly.py` line 51 at r5](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-NdkBwj_42MEdj3yE9-B:-NdkBwj_42MEdj3yE9-C:b5tl1gy) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ccd8c66bc9ff29ccf0ffac4ff4079f07eaf56f5f/full_assembly.py#L51)):*
> ```Python
> YAML_FILE = args.config
> yaml_content = Classifier.read_config(YAML_FILE)
> config_fields = list(yaml_content["table_cond_formatting_rules"].keys())
> ```

can all of this go into `main()` or separate functions for loading files



_Code quote:_
```Python
args = parser.parse_args()
# List of samples
sample_sheet = args.samplesheet
MULTIQC_DATA_FILE = args.data
with open(MULTIQC_DATA_FILE, 'r', encoding='UTF-8') as file:
    multiqc_data = json.load(file)
sample_lists = Classifier.get_sample_lists(multiqc_data, sample_sheet)

# List of config_fields
YAML_FILE = args.config
yaml_content = Classifier.read_config(YAML_FILE)
config_fields = list(yaml_content["table_cond_formatting_rules"].keys())
```


<!-- Sent from Reviewable.io -->

mattgarner committed:
Delete testfiles/SampleSheet.csv

Commit:
Minor commit to save output to qc_report.json
	modified:   .gitignore
	modified:   bin/utils.py
	modified:   full_assembly.py

Commit:
Merge branch 'scripts_and_data' of https://github.com/eastgenomics/QC_Classifier into scripts_and_data

jethror1 reviewed (commented):
> Reviewed 2 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 8 of 10 files reviewed, 3 unresolved discussions (waiting on @Arun-bioinformatics)



<!-- Sent from Reviewable.io -->

Arun-bioinformatics reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: 8 of 10 files reviewed, 3 unresolved discussions (waiting on @jethror1)

___
*[`full_assembly.py` line 40 at r5](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-NdkC56kEiuDoQN0YhnO:-Ne80xoJ7jsILUw6J1sw:b-uheco7) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ccd8c66bc9ff29ccf0ffac4ff4079f07eaf56f5f/full_assembly.py#L40)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

please can this be in its own function too please and not global
</blockquote></details>

Done, implemented in latest branch
___
*[`full_assembly.py` line 51 at r5](https://reviewable.io/reviews/eastgenomics/qc_classifier/2#-NdkBwj_42MEdj3yE9-B:-Ne810PSBvS0xGYPEBpp:b-lkw74b) ([raw file](https://github.com/eastgenomics/qc_classifier/blob/ccd8c66bc9ff29ccf0ffac4ff4079f07eaf56f5f/full_assembly.py#L51)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

can all of this go into `main()` or separate functions for loading files
</blockquote></details>

Done, implemented in latest branch.


<!-- Sent from Reviewable.io -->

jethror1 reviewed (approved):
> Reviewed 1 of 1 files at r6, 1 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/qc_classifier/2)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Arun-bioinformatics)



<!-- Sent from Reviewable.io -->
